File: javascript/017/main.js
/*
 * ProjectEuler 017
 * JavaScript / main.js
 * copyright (c) 2015 Susisu
 *
 * Node.js v4.2.1
 * `node main.js`
 */

"use strict";

function main() {
    var sum = 0;
    for (var i = 1; i <= 1000; i++) {
        sum += numberToWord(i).length;
    }
    console.log(sum);
}

var table = [
    "zero",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
    "ten",
    "eleven",
    "twelve",
    "thirteen",
    "fourteen",
    "fifteen",
    "sixteen",
    "seventeen",
    "eighteen",
    "nineteen",
    "twenty"
];
table[30]   = "thrity";
table[40]   = "forty";
table[50]   = "fifty";
table[60]   = "sixty";
table[70]   = "seventy";
table[80]   = "eighty";
table[90]   = "ninety";
table[100]  = "hundred";

function numberToWord(n) {
    if (n == 0) {
        return "";
    }
    else if (n <= 20) {
        return table[n];
    }
    else if (n < 100) {
        return table[Math.floor(n / 10) * 10] + numberToWord(n % 10);
    }
    else if (n < 1000) {
        return table[Math.floor(n / 100)] + table[100] + (n % 100 > 0 ? "and" + numberToWord(n % 100) : "");
    }
    else if (n == 1000) {
        return "onethousand";
    }
    else {
        return "";
    }
}

main();

File: ocaml/017/main.ml
(*
    ProjectEuler 017
    OCaml / main.ml
    copyright (c) 2015 Susisu

    OCaml 4.02.3
    `ocamlfind ocamlopt -o bin/main -linkpkg -package core -thread nums.cmxa main.ml`
*)

open Core;;

let words_under_20 = [|
        "zero";
        "one";
        "two";
        "three";
        "four";
        "five";
        "six";
        "seven";
        "eight";
        "nine";
        "ten";
        "eleven";
        "twelve";
        "thirteen";
        "fourteen";
        "fifteen";
        "sixteen";
        "seventeen";
        "eighteen";
        "nineteen"
    |]

let words_multiple_10 = [|
        "zero";
        "ten";
        "twenty";
        "thirty";
        "forty";
        "fifty";
        "sixty";
        "seventy";
        "eighty";
        "ninety"
    |]

let rec word_of_number ?(conj=false) n =
    if n <= 0    then ""
    else
        let c    = if conj then "and" else "" in
        let word =
                 if n < 20    then Array.get words_under_20 n
            else if n < 100   then Array.get words_multiple_10 (n / 10) ^ word_of_number (n mod 10) 
            else if n < 1000  then Array.get words_under_20 (n / 100) ^ "hundred" ^ word_of_number ~conj:true (n mod 100)
            else if n == 1000 then "onethousand"
            else ""
        in c ^ word

let main _ =
    let xs    = Core_list.range ~stop:`inclusive 1 1000 in
    let words = List.map word_of_number xs in
    let wls   = List.map String.length words in
    print_int (List.fold_left (+) 0 wls);
    print_newline ()

let _ = main ()
